File: plugins/plugin-evm/rust/src/actions/bridge.rs
#![allow(missing_docs)]

use alloy::{
    hex,
    network::TransactionBuilder,
    primitives::{Address, Bytes, U256},
    providers::Provider,
    rpc::types::TransactionRequest,
};
use serde::{Deserialize, Serialize};
use std::sync::Arc;
use std::time::Duration;

use crate::constants::{
    BRIDGE_POLL_INTERVAL_SECS, DEFAULT_SLIPPAGE_PERCENT, LIFI_API_URL, MAX_BRIDGE_POLL_ATTEMPTS,
    MAX_PRICE_IMPACT,
};
use crate::error::{EVMError, EVMErrorCode, EVMResult};
use crate::providers::WalletProvider;
use crate::types::{SupportedChain, Transaction};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BridgeParams {
    pub from_chain: SupportedChain,
    pub to_chain: SupportedChain,
    pub from_token: Address,
    pub to_token: Address,
    pub amount: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub to_address: Option<Address>,
}

impl BridgeParams {
    #[must_use]
    pub fn new(
        from_chain: SupportedChain,
        to_chain: SupportedChain,
        from_token: Address,
        to_token: Address,
        amount: impl Into<String>,
    ) -> Self {
        Self {
            from_chain,
            to_chain,
            from_token,
            to_token,
            amount: amount.into(),
            to_address: None,
        }
    }

    #[must_use]
    pub fn with_recipient(mut self, address: Address) -> Self {
        self.to_address = Some(address);
        self
    }

    pub fn validate(&self) -> EVMResult<()> {
        if self.amount.is_empty() {
            return Err(EVMError::invalid_params("Amount is required"));
        }

        if self.from_chain == self.to_chain {
            return Err(EVMError::invalid_params(
                "Source and destination chains must be different",
            ));
        }

        Ok(())
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BridgeStatus {
    pub status: BridgeStatusType,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub substatus: Option<String>,
    pub source_tx_hash: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub dest_tx_hash: Option<String>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "UPPERCASE")]
pub enum BridgeStatusType {
    Pending,
    Done,
    Failed,
}

#[derive(Debug, Deserialize)]
struct LifiRoutesResponse {
    routes: Vec<LifiRoute>,
}

#[derive(Debug, Deserialize)]
struct LifiRoute {
    steps: Vec<LifiStep>,
    #[serde(rename = "fromChainId")]
    from_chain_id: u64,
    #[serde(rename = "toChainId")]
    to_chain_id: u64,
}

#[derive(Debug, Deserialize)]
struct LifiStep {
    tool: String,
    #[serde(rename = "transactionRequest")]
    transaction_request: Option<LifiTransactionRequest>,
}

#[derive(Debug, Deserialize)]
struct LifiTransactionRequest {
    to: String,
    value: String,
    data: String,
    #[serde(rename = "gasLimit")]
    gas_limit: Option<String>,
}

#[derive(Debug, Deserialize)]
struct LifiStatusResponse {
    status: String,
    substatus: Option<String>,
    #[serde(rename = "receiving")]
    receiving: Option<LifiReceiving>,
}

#[derive(Debug, Deserialize)]
struct LifiReceiving {
    #[serde(rename = "txHash")]
    tx_hash: Option<String>,
}

pub struct BridgeAction {
    provider: Arc<WalletProvider>,
    http_client: reqwest::Client,
}

impl BridgeAction {
    /// elizaOS Action name (TS parity).
    pub const NAME: &'static str = "EVM_BRIDGE_TOKENS";

    #[must_use]
    pub fn new(provider: Arc<WalletProvider>) -> Self {
        Self {
            provider,
            http_client: reqwest::Client::builder()
                .timeout(Duration::from_secs(30))
                .build()
                .unwrap_or_default(),
        }
    }

    async fn get_routes(&self, params: &BridgeParams) -> EVMResult<LifiRoutesResponse> {
        let from_address = self.provider.address();
        let to_address = params.to_address.unwrap_or(from_address);

        let url = format!(
            "{}/routes?fromChainId={}&toChainId={}&fromTokenAddress={}&toTokenAddress={}&fromAmount={}&fromAddress={}&toAddress={}&options={{\"slippage\":{},\"maxPriceImpact\":{}}}",
            LIFI_API_URL,
            params.from_chain.chain_id(),
            params.to_chain.chain_id(),
            params.from_token,
            params.to_token,
            params.amount,
            from_address,
            to_address,
            DEFAULT_SLIPPAGE_PERCENT,
            MAX_PRICE_IMPACT,
        );

        let response = self
            .http_client
            .get(&url)
            .header("Accept", "application/json")
            .send()
            .await?;

        if !response.status().is_success() {
            let status = response.status();
            let body = response.text().await.unwrap_or_default();
            return Err(EVMError::new(
                EVMErrorCode::RouteNotFound,
                format!("LiFi API error ({}): {}", status, body),
            ));
        }

        response.json().await.map_err(Into::into)
    }

    async fn poll_status(
        &self,
        tx_hash: &str,
        from_chain_id: u64,
        to_chain_id: u64,
        tool: &str,
    ) -> EVMResult<BridgeStatus> {
        for attempt in 1..=MAX_BRIDGE_POLL_ATTEMPTS {
            tokio::time::sleep(Duration::from_secs(BRIDGE_POLL_INTERVAL_SECS)).await;

            let url = format!(
                "{}/status?txHash={}&fromChain={}&toChain={}&bridge={}",
                LIFI_API_URL, tx_hash, from_chain_id, to_chain_id, tool
            );

            let response = self
                .http_client
                .get(&url)
                .header("Accept", "application/json")
                .send()
                .await;

            match response {
                Ok(resp) if resp.status().is_success() => {
                    let status_response: LifiStatusResponse = resp.json().await?;

                    let status_type = match status_response.status.as_str() {
                        "DONE" => BridgeStatusType::Done,
                        "FAILED" => BridgeStatusType::Failed,
                        _ => BridgeStatusType::Pending,
                    };

                    let dest_tx_hash = status_response.receiving.and_then(|r| r.tx_hash);

                    if status_type != BridgeStatusType::Pending {
                        return Ok(BridgeStatus {
                            status: status_type,
                            substatus: status_response.substatus,
                            source_tx_hash: tx_hash.to_string(),
                            dest_tx_hash,
                        });
                    }

                    tracing::debug!(
                        "Bridge poll attempt {}/{}: {}",
                        attempt,
                        MAX_BRIDGE_POLL_ATTEMPTS,
                        status_response.status
                    );
                }
                Ok(resp) => {
                    tracing::warn!("Status check failed with code: {}", resp.status());
                }
                Err(e) => {
                    tracing::warn!("Status check error: {}", e);
                }
            }
        }

        Ok(BridgeStatus {
            status: BridgeStatusType::Pending,
            substatus: Some("Status polling timed out".to_string()),
            source_tx_hash: tx_hash.to_string(),
            dest_tx_hash: None,
        })
    }

    pub async fn execute(&self, params: BridgeParams) -> EVMResult<(Transaction, BridgeStatus)> {
        params.validate()?;

        self.provider.provider(params.from_chain)?;
        self.provider.provider(params.to_chain)?;

        let chain_provider = self.provider.provider(params.from_chain)?;
        let chain_id = params.from_chain.chain_id();
        let routes = self.get_routes(&params).await?;

        let route = routes
            .routes
            .first()
            .ok_or_else(|| EVMError::new(EVMErrorCode::RouteNotFound, "No bridge routes found"))?;

        let step = route
            .steps
            .first()
            .ok_or_else(|| EVMError::new(EVMErrorCode::RouteNotFound, "No bridge steps found"))?;

        let tx_request = step.transaction_request.as_ref().ok_or_else(|| {
            EVMError::new(
                EVMErrorCode::RouteNotFound,
                "No transaction request in route",
            )
        })?;

        let to: Address = tx_request
            .to
            .parse()
            .map_err(|_| EVMError::invalid_params("Invalid to address in route"))?;

        let value = U256::from_str_radix(tx_request.value.trim_start_matches("0x"), 16)
            .unwrap_or(U256::ZERO);

        let data = Bytes::from(
            hex::decode(tx_request.data.trim_start_matches("0x"))
                .map_err(|e| EVMError::invalid_params(format!("Invalid data in route: {e}")))?,
        );

        let gas_limit: Option<u64> = tx_request.gas_limit.as_ref().and_then(|g| g.parse().ok());

        let mut tx = TransactionRequest::default()
            .with_to(to)
            .with_value(value)
            .with_input(data.clone());

        if let Some(limit) = gas_limit {
            tx = tx.with_gas_limit(limit);
        }

        let pending = chain_provider.send_transaction(tx).await.map_err(|e| {
            EVMError::new(
                EVMErrorCode::TransactionFailed,
                format!("Failed to send bridge transaction: {e}"),
            )
        })?;

        let tx_hash = *pending.tx_hash();
        let tx_hash_str = format!("{tx_hash:#x}");
        let receipt = pending.get_receipt().await.map_err(|e| {
            EVMError::new(
                EVMErrorCode::TransactionFailed,
                format!("Failed to get receipt: {e}"),
            )
        })?;

        if !receipt.status() {
            return Err(EVMError::new(
                EVMErrorCode::ContractRevert,
                format!("Bridge transaction reverted: {tx_hash}"),
            ));
        }

        tracing::info!("Bridge source transaction confirmed: {}", tx_hash);

        let bridge_status = self
            .poll_status(
                &tx_hash_str,
                route.from_chain_id,
                route.to_chain_id,
                &step.tool,
            )
            .await?;

        let transaction = Transaction {
            hash: tx_hash,
            from: self.provider.address(),
            to,
            value,
            data: Some(data),
            chain_id: Some(chain_id),
        };

        Ok((transaction, bridge_status))
    }

    pub async fn check_status(
        &self,
        tx_hash: &str,
        from_chain: SupportedChain,
        to_chain: SupportedChain,
        tool: &str,
    ) -> EVMResult<BridgeStatus> {
        let url = format!(
            "{}/status?txHash={}&fromChain={}&toChain={}&bridge={}",
            LIFI_API_URL,
            tx_hash,
            from_chain.chain_id(),
            to_chain.chain_id(),
            tool
        );

        let response = self
            .http_client
            .get(&url)
            .header("Accept", "application/json")
            .send()
            .await?;

        if !response.status().is_success() {
            return Err(EVMError::network_error("Failed to get bridge status"));
        }

        let status_response: LifiStatusResponse = response.json().await?;

        let status_type = match status_response.status.as_str() {
            "DONE" => BridgeStatusType::Done,
            "FAILED" => BridgeStatusType::Failed,
            _ => BridgeStatusType::Pending,
        };

        Ok(BridgeStatus {
            status: status_type,
            substatus: status_response.substatus,
            source_tx_hash: tx_hash.to_string(),
            dest_tx_hash: status_response.receiving.and_then(|r| r.tx_hash),
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use alloy::primitives::address;

    #[test]
    fn test_bridge_params_validation() {
        let native = Address::ZERO;
        let params = BridgeParams::new(
            SupportedChain::Mainnet,
            SupportedChain::Base,
            native,
            native,
            "1000000000000000000",
        );

        assert!(params.validate().is_ok());
    }

    #[test]
    fn test_same_chain_error() {
        let native = Address::ZERO;
        let params = BridgeParams::new(
            SupportedChain::Mainnet,
            SupportedChain::Mainnet,
            native,
            native,
            "1000000000000000000",
        );

        assert!(params.validate().is_err());
    }

    #[test]
    fn test_bridge_with_recipient() {
        let native = Address::ZERO;
        let recipient = address!("d8dA6BF26964aF9D7eEd9e03E53415D37aA96045");
        let params = BridgeParams::new(
            SupportedChain::Mainnet,
            SupportedChain::Base,
            native,
            native,
            "1000000000000000000",
        )
        .with_recipient(recipient);

        assert_eq!(params.to_address, Some(recipient));
    }
}

File: plugins/plugin-evm/typescript/actions/bridge.ts
import type { ActionResult, HandlerCallback, IAgentRuntime, Memory, State } from "@elizaos/core";
import { composePromptFromState, logger, ModelType, parseKeyValueXml } from "@elizaos/core";
import {
  createConfig,
  EVM,
  type ExecutionOptions,
  type ExtendedChain,
  executeRoute,
  getRoutes,
  getStatus,
  getToken,
  type RouteExtended,
  resumeRoute,
} from "@lifi/sdk";
import { type Address, parseAbi, parseUnits } from "viem";
import {
  BRIDGE_POLL_INTERVAL_MS,
  DEFAULT_SLIPPAGE_PERCENT,
  MAX_BRIDGE_POLL_ATTEMPTS,
  MAX_PRICE_IMPACT,
  NATIVE_TOKEN_ADDRESS,
} from "../constants";
import { requireActionSpec } from "../generated/specs/spec-helpers";
import { initWalletProvider, type WalletProvider } from "../providers/wallet";
import { bridgeTemplate } from "../templates";
import {
  type BridgeParams,
  EVMError,
  EVMErrorCode,
  parseBridgeParams,
  type SupportedChain,
  type Transaction,
} from "../types";

export { bridgeTemplate };

type LiFiGetWalletClient = Parameters<typeof EVM>[0]["getWalletClient"];
type LiFiSwitchChain = Parameters<typeof EVM>[0]["switchChain"];

function createLiFiGetWalletClientAdapter(
  walletProvider: WalletProvider,
  getFirstChain: () => string
): LiFiGetWalletClient {
  return (async () => {
    const firstChain = getFirstChain();
    return walletProvider.getWalletClient(firstChain as SupportedChain);
  }) as LiFiGetWalletClient;
}

function createLiFiSwitchChainAdapter(
  walletProvider: WalletProvider,
  getChainNameById: (chainId: number) => string
): LiFiSwitchChain {
  return (async (chainId: number) => {
    logger.debug(`LiFi requesting chain switch to ${chainId}...`);
    const chainName = getChainNameById(chainId);
    return walletProvider.getWalletClient(chainName as SupportedChain);
  }) as LiFiSwitchChain;
}

function createExecutionSwitchChainHookAdapter(
  walletProvider: WalletProvider,
  getChainNameById: (chainId: number) => string
): ExecutionOptions["switchChainHook"] {
  return (async (chainId: number) => {
    logger.debug(`Switching to chain ${chainId}...`);
    const chainName = getChainNameById(chainId);
    return walletProvider.getWalletClient(chainName as SupportedChain);
  }) as ExecutionOptions["switchChainHook"];
}

interface BridgeExecutionStatus {
  readonly route: RouteExtended;
  readonly isComplete: boolean;
  readonly error?: string;
  readonly transactionHashes: readonly string[];
  readonly currentStep: number;
  readonly totalSteps: number;
}

export class BridgeAction {
  private readonly activeRoutes: Map<string, BridgeExecutionStatus> = new Map();

  constructor(private readonly walletProvider: WalletProvider) {
    const evmProvider = EVM({
      getWalletClient: createLiFiGetWalletClientAdapter(
        this.walletProvider,
        () => Object.keys(this.walletProvider.chains)[0]
      ),
      switchChain: createLiFiSwitchChainAdapter(this.walletProvider, (chainId: number) =>
        this.getChainNameById(chainId)
      ),
    });

    createConfig({
      integrator: "eliza-agent",
      providers: [evmProvider],
      chains: Object.values(this.walletProvider.chains).map((config) => ({
        id: config.id,
        name: config.name,
        key: config.name.toLowerCase(),
        chainType: "EVM",
        nativeToken: {
          ...config.nativeCurrency,
          chainId: config.id,
          address: NATIVE_TOKEN_ADDRESS,
          coinKey: config.nativeCurrency.symbol,
        },
        metamask: {
          chainId: `0x${config.id.toString(16)}`,
          chainName: config.name,
          nativeCurrency: config.nativeCurrency,
          rpcUrls: [config.rpcUrls.default.http[0]],
          blockExplorerUrls: config.blockExplorers?.default?.url
            ? [config.blockExplorers.default.url]
            : [],
        },
        diamondAddress: NATIVE_TOKEN_ADDRESS,
        coin: config.nativeCurrency.symbol,
        mainnet: true,
      })) as ExtendedChain[],
      routeOptions: {
        maxPriceImpact: MAX_PRICE_IMPACT,
        slippage: DEFAULT_SLIPPAGE_PERCENT,
      },
    });
  }

  private getChainNameById(chainId: number): string {
    const chain = Object.entries(this.walletProvider.chains).find(
      ([_, config]) => config.id === chainId
    );
    if (!chain) {
      throw new EVMError(EVMErrorCode.CHAIN_NOT_CONFIGURED, `Chain with ID ${chainId} not found`);
    }
    return chain[0];
  }

  private async resolveTokenAddress(
    tokenSymbolOrAddress: string,
    chainId: number
  ): Promise<string> {
    if (tokenSymbolOrAddress.startsWith("0x") && tokenSymbolOrAddress.length === 42) {
      return tokenSymbolOrAddress;
    }

    if (tokenSymbolOrAddress === NATIVE_TOKEN_ADDRESS) {
      return tokenSymbolOrAddress;
    }

    const token = await getToken(chainId, tokenSymbolOrAddress);
    return token.address;
  }

  private async getTokenDecimals(tokenAddress: string, chainName: string): Promise<number> {
    const chainConfig = this.walletProvider.getChainConfigs(chainName as SupportedChain);

    if (
      tokenAddress === NATIVE_TOKEN_ADDRESS ||
      tokenAddress.toUpperCase() === chainConfig.nativeCurrency.symbol.toUpperCase()
    ) {
      return chainConfig.nativeCurrency.decimals;
    }

    const decimalsAbi = parseAbi(["function decimals() view returns (uint8)"]);

    const publicClient = this.walletProvider.getPublicClient(chainName as SupportedChain);
    // @ts-expect-error - viem type narrowing issue with readContract parameters
    const decimals = (await publicClient.readContract({
      address: tokenAddress as Address,
      abi: decimalsAbi,
      functionName: "decimals",
    })) as number;
    return Number(decimals);
  }

  private createExecutionOptions(routeId: string): ExecutionOptions {
    return {
      updateTransactionRequestHook: async (txRequest) => {
        if (txRequest.gas) {
          txRequest.gas = (BigInt(txRequest.gas) * BigInt(110)) / BigInt(100);
        }
        if (txRequest.gasPrice) {
          txRequest.gasPrice = (BigInt(txRequest.gasPrice) * BigInt(105)) / BigInt(100);
        }
        return txRequest;
      },

      acceptExchangeRateUpdateHook: async (params: {
        toToken: { decimals: number; symbol: string };
        oldToAmount: string;
        newToAmount: string;
      }) => {
        const priceChange =
          ((Number(params.newToAmount) - Number(params.oldToAmount)) / Number(params.oldToAmount)) *
          100;

        logger.debug(`Exchange rate change: ${priceChange.toFixed(2)}%`);
        return Math.abs(priceChange) < 5;
      },

      updateRouteHook: (updatedRoute: RouteExtended) => {
        this.updateRouteStatus(routeId, updatedRoute);
      },

      switchChainHook: createExecutionSwitchChainHookAdapter(
        this.walletProvider,
        (chainId: number) => this.getChainNameById(chainId)
      ),

      executeInBackground: false,
      disableMessageSigning: false,
    };
  }

  private updateRouteStatus(routeId: string, route: RouteExtended): BridgeExecutionStatus {
    const transactionHashes: string[] = [];
    let currentStep = 0;
    let isComplete = false;
    let error: string | undefined;

    route.steps.forEach((step, stepIndex) => {
      const stepExecution = step.execution;
      if (stepExecution?.process) {
        stepExecution.process.forEach((process) => {
          if (process.txHash) {
            transactionHashes.push(process.txHash);
          }
          if (process.status === "DONE") {
            currentStep = Math.max(currentStep, stepIndex + 1);
          }
          if (process.status === "FAILED") {
            error = `Step ${stepIndex + 1} failed: ${process.error ?? "Unknown error"}`;
          }
        });
      }
    });

    isComplete = currentStep === route.steps.length && !error;

    const status: BridgeExecutionStatus = {
      route,
      isComplete,
      error,
      transactionHashes,
      currentStep,
      totalSteps: route.steps.length,
    };

    this.activeRoutes.set(routeId, status);
    return status;
  }

  private async pollBridgeStatus(
    txHash: string,
    fromChainId: number,
    toChainId: number,
    tool: string,
    routeId: string
  ): Promise<BridgeExecutionStatus> {
    for (let attempt = 1; attempt <= MAX_BRIDGE_POLL_ATTEMPTS; attempt++) {
      await new Promise((resolve) => setTimeout(resolve, BRIDGE_POLL_INTERVAL_MS));

      try {
        const status = await getStatus({
          txHash,
          fromChain: fromChainId,
          toChain: toChainId,
          bridge: tool,
        });

        logger.debug(`Poll attempt ${attempt}/${MAX_BRIDGE_POLL_ATTEMPTS}: ${status.status}`);

        const routeStatus = this.activeRoutes.get(routeId);
        if (!routeStatus) {
          throw new EVMError(EVMErrorCode.INVALID_PARAMS, `Route ${routeId} not found`);
        }

        let isComplete = false;
        let error: string | undefined;

        if (status.status === "DONE") {
          isComplete = true;
          logger.debug("Bridge completed successfully!");
        } else if (status.status === "FAILED") {
          error = `Bridge failed: ${status.substatus ?? "Unknown error"}`;
        }

        const updatedStatus: BridgeExecutionStatus = {
          ...routeStatus,
          isComplete,
          error,
          currentStep: isComplete ? routeStatus.totalSteps : routeStatus.currentStep,
        };

        this.activeRoutes.set(routeId, updatedStatus);

        if (isComplete || error) {
          return updatedStatus;
        }
      } catch (statusError) {
        logger.warn(`Status check attempt ${attempt} failed:`, statusError);
      }
    }

    const routeStatus = this.activeRoutes.get(routeId);
    if (routeStatus) {
      const timeoutStatus: BridgeExecutionStatus = {
        ...routeStatus,
        error: `Bridge status polling timed out after ${(MAX_BRIDGE_POLL_ATTEMPTS * BRIDGE_POLL_INTERVAL_MS) / 1000}s`,
      };
      this.activeRoutes.set(routeId, timeoutStatus);
      return timeoutStatus;
    }

    throw new EVMError(EVMErrorCode.NETWORK_ERROR, "Route status polling failed");
  }

  async bridge(params: BridgeParams): Promise<Transaction> {
    // Validate inputs early to fail fast
    const amount = parseFloat(params.amount);
    if (Number.isNaN(amount) || amount <= 0) {
      throw new EVMError(EVMErrorCode.INVALID_PARAMS, "Amount must be a positive number");
    }

    if (params.fromChain === params.toChain) {
      throw new EVMError(
        EVMErrorCode.INVALID_PARAMS,
        "Source and destination chains must be different for bridging"
      );
    }

    if (
      params.toAddress &&
      (!params.toAddress.startsWith("0x") || params.toAddress.length !== 42)
    ) {
      throw new EVMError(
        EVMErrorCode.INVALID_PARAMS,
        `Invalid recipient address: ${params.toAddress}`
      );
    }

    const walletClient = this.walletProvider.getWalletClient(params.fromChain);
    const [fromAddress] = await walletClient.getAddresses();

    logger.debug(`Bridge: ${params.fromChain} → ${params.toChain}`);
    logger.debug(`Amount: ${params.amount}`);

    const fromChainConfig = this.walletProvider.getChainConfigs(params.fromChain);
    const toChainConfig = this.walletProvider.getChainConfigs(params.toChain);

    const resolvedFromToken = await this.resolveTokenAddress(params.fromToken, fromChainConfig.id);
    const resolvedToToken = await this.resolveTokenAddress(params.toToken, toChainConfig.id);

    const fromTokenDecimals = await this.getTokenDecimals(resolvedFromToken, params.fromChain);

    const fromAmountParsed = parseUnits(params.amount, fromTokenDecimals);

    const routesResult = await getRoutes({
      fromChainId: fromChainConfig.id,
      toChainId: toChainConfig.id,
      fromTokenAddress: resolvedFromToken,
      toTokenAddress: resolvedToToken,
      fromAmount: fromAmountParsed.toString(),
      fromAddress,
      toAddress: params.toAddress ?? fromAddress,
      options: {
        order: "RECOMMENDED",
        slippage: DEFAULT_SLIPPAGE_PERCENT,
        maxPriceImpact: MAX_PRICE_IMPACT,
        allowSwitchChain: true,
      },
    });

    if (!routesResult.routes.length) {
      throw new EVMError(
        EVMErrorCode.INVALID_PARAMS,
        `No bridge routes found for ${params.fromToken} on ${params.fromChain} to ${params.toToken} on ${params.toChain}`
      );
    }

    const selectedRoute = routesResult.routes[0];
    const routeId = `bridge_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;

    logger.debug(`Selected route ${routeId}`);
    logger.debug(`Steps: ${selectedRoute.steps.length}`);

    try {
      const executionOptions = this.createExecutionOptions(routeId);
      const executedRoute = await executeRoute(selectedRoute, executionOptions);

      const sourceSteps = executedRoute.steps.filter((step) =>
        step.execution?.process?.some((p) => p.txHash)
      );

      if (!sourceSteps.length) {
        throw new EVMError(EVMErrorCode.NETWORK_ERROR, "No transaction hashes found");
      }

      const mainTxHash = sourceSteps[0]?.execution?.process?.find((p) => p.txHash)?.txHash;

      if (!mainTxHash) {
        throw new EVMError(EVMErrorCode.NETWORK_ERROR, "No transaction hash found");
      }

      logger.debug(`Source transaction: ${mainTxHash}`);

      const bridgeTool = selectedRoute.steps[0].tool;
      const finalStatus = await this.pollBridgeStatus(
        mainTxHash,
        fromChainConfig.id,
        toChainConfig.id,
        bridgeTool,
        routeId
      );

      if (finalStatus.error) {
        throw new EVMError(EVMErrorCode.CONTRACT_REVERT, finalStatus.error);
      }

      logger.debug("Bridge initiated successfully!");

      return {
        hash: mainTxHash as `0x${string}`,
        from: fromAddress,
        to: (params.toAddress ?? fromAddress) as `0x${string}`,
        value: fromAmountParsed,
        chainId: toChainConfig.id,
      };
    } finally {
      this.activeRoutes.delete(routeId);
    }
  }

  async getTransactionStatus(txHash: string, fromChainId: number, toChainId: number, tool: string) {
    return await getStatus({
      txHash,
      fromChain: fromChainId,
      toChain: toChainId,
      bridge: tool,
    });
  }

  async resumeBridge(route: RouteExtended) {
    const routeId = `resume_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;
    const executionOptions = this.createExecutionOptions(routeId);

    logger.debug("Resuming bridge operation...");

    try {
      return await resumeRoute(route, executionOptions);
    } finally {
      this.activeRoutes.delete(routeId);
    }
  }
}

async function buildBridgeDetails(
  state: State,
  runtime: IAgentRuntime,
  wp: WalletProvider
): Promise<BridgeParams> {
  const chains = wp.getSupportedChains();
  const balances = await wp.getWalletBalances();

  state.supportedChains = chains.join(" | ");
  state.chainBalances = Object.entries(balances)
    .map(([chain, balance]) => {
      const chainConfig = wp.getChainConfigs(chain as SupportedChain);
      return `${chain}: ${balance} ${chainConfig.nativeCurrency.symbol}`;
    })
    .join(", ");

  const bridgeContext = composePromptFromState({
    state,
    template: bridgeTemplate,
  });

  const xmlResponse = await runtime.useModel(ModelType.TEXT_LARGE, {
    prompt: bridgeContext,
  });

  const content = parseKeyValueXml(xmlResponse);

  if (!content) {
    throw new EVMError(EVMErrorCode.INVALID_PARAMS, "Failed to parse bridge details from LLM");
  }

  const rawParams = {
    fromChain: String(content.fromChain ?? "").toLowerCase(),
    toChain: String(content.toChain ?? "").toLowerCase(),
    fromToken: String(content.token ?? ""),
    toToken: String(content.token ?? ""),
    amount: String(content.amount ?? ""),
    toAddress: content.toAddress ? String(content.toAddress) : undefined,
  };

  const bridgeOptions = parseBridgeParams(rawParams);

  if (!wp.chains[bridgeOptions.fromChain]) {
    throw new EVMError(
      EVMErrorCode.CHAIN_NOT_CONFIGURED,
      `Source chain ${bridgeOptions.fromChain} not configured. Available: ${chains.join(", ")}`
    );
  }

  if (!wp.chains[bridgeOptions.toChain]) {
    throw new EVMError(
      EVMErrorCode.CHAIN_NOT_CONFIGURED,
      `Destination chain ${bridgeOptions.toChain} not configured. Available: ${chains.join(", ")}`
    );
  }

  return bridgeOptions;
}

const spec = requireActionSpec("BRIDGE");

export const bridgeAction = {
  name: spec.name,
  description: spec.description,

  handler: async (
    runtime: IAgentRuntime,
    message: Memory,
    state?: State,
    _options?: Record<string, unknown>,
    callback?: HandlerCallback
  ): Promise<ActionResult> => {
    const walletProvider = await initWalletProvider(runtime);
    const action = new BridgeAction(walletProvider);

    if (!state) {
      state = await runtime.composeState(message, ["RECENT_MESSAGES"], true);
    }

    const bridgeOptions = await buildBridgeDetails(state, runtime, walletProvider);

    const bridgeResp = await action.bridge(bridgeOptions);

    const successText = `✅ Successfully bridged ${bridgeOptions.amount} tokens from ${bridgeOptions.fromChain} to ${bridgeOptions.toChain}\n\nTransaction Hash: ${bridgeResp.hash}`;

    await runtime.createMemory(
      {
        entityId: message.agentId ?? runtime.agentId,
        roomId: message.roomId,
        agentId: message.agentId ?? runtime.agentId,
        content: {
          text: successText,
          action: ["EVM_BRIDGE_TOKENS"],
        },
      },
      "messages"
    );

    if (callback) {
      callback({
        text: successText,
        content: {
          success: true,
          hash: bridgeResp.hash,
          recipient: bridgeResp.to,
          fromChain: bridgeOptions.fromChain,
          toChain: bridgeOptions.toChain,
          amount: bridgeOptions.amount,
        },
      });
    }

    return {
      success: true,
      text: successText,
      values: {
        bridgeSucceeded: true,
      },
      data: {
        actionName: "EVM_BRIDGE_TOKENS",
        transactionHash: bridgeResp.hash,
        fromChain: bridgeOptions.fromChain,
        toChain: bridgeOptions.toChain,
        token: bridgeOptions.fromToken,
        amount: bridgeOptions.amount,
        recipient: bridgeResp.to,
      },
    };
  },

  template: bridgeTemplate,

  validate: async (runtime: IAgentRuntime): Promise<boolean> => {
    const privateKey = runtime.getSetting("EVM_PRIVATE_KEY");
    return typeof privateKey === "string" && privateKey.startsWith("0x");
  },

  examples: [
    [
      {
        name: "user",
        user: "user",
        content: {
          text: "Bridge 1 ETH from Ethereum to Base",
          action: "CROSS_CHAIN_TRANSFER",
        },
      },
    ],
  ],

  similes: spec.similes ? [...spec.similes] : [],
};

export async function checkBridgeStatus(
  txHash: string,
  fromChainId: number,
  toChainId: number,
  tool: string = "stargateV2Bus"
) {
  const status = await getStatus({
    txHash,
    fromChain: fromChainId,
    toChain: toChainId,
    bridge: tool,
  });

  logger.debug(
    `Bridge Status: ${status.status}${status.substatus ? ` (${status.substatus})` : ""}`
  );

  return {
    status: status.status,
    substatus: status.substatus,
    isComplete: status.status === "DONE",
    isFailed: status.status === "FAILED",
    isPending: status.status === "PENDING",
    error: status.status === "FAILED" ? status.substatus : undefined,
  };
}
